### PR TITLE
[Triton] Add end-to-end example for dispatching Triton functions

### PR DESCRIPTION
**Pull request by ezhulenev**
Initial implementation of the First class triton integration: https://github.com/openxla/openxla-nvgpu/issues/54

Requires Triton + patches from https://github.com/ezhulenev/triton/commits/openxla-triton

```
git submodule update --remote third_party/triton
```

Or a PR to be landed in Triton: https://github.com/openai/triton/pull/1668

Run tests:
```
ctest --test-dir build -R TritonFlow
ctest --test-dir build -R triton
```

### CODE REVIEW

**Review by frgossen — APPROVED** [review]
Super cool!

File: compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Conversion/ConvertTritonToFlowDispatch.cpp [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+}
+
+//===----------------------------------------------------------------------===//
+// Compilation pipeline from Triton IR to LLVM IR
+//===----------------------------------------------------------------------===//
+
+void buildTritonCompilationPipeline(mlir::OpPassManager &pm,
+                                    const TritonOptions &opts) {
+  // Based on optimize_ttir() in
+  // @triton//:python/triton/compiler/compiler.py
```
Comment (stellaraccident) [inline compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Conversion/ConvertTritonToFlowDispatch.cpp:94]: It'd be really good to have a better source of truth on this vs a bunch of freestanding passes that we maintain.

File: compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Conversion/ConvertTritonToFlowDispatch.cpp [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    SymbolTable sym_table(op->getParentOfType<ModuleOp>());
+    auto fn = sym_table.lookupNearestSymbolFrom<FuncOp>(op, op.getCalleeAttr());
+
+    // Get a PTX module from the Triton callee.
+    FailureOr<std::string> ptx = compileTritonFunctionToPTX(fn, opts);
+    if (failed(ptx))
+      return rewriter.notifyMatchFailure(
+          op, "Failed to lower Triton function to LLVM IR");
+
+    // Write PTX to a temp file, so that we can constuct an HAL executable.
```
Comment (stellaraccident) [inline compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Conversion/ConvertTritonToFlowDispatch.cpp:212]: Can you add a TODO to figure out how to do this in-memory?

**Review by stellaraccident — APPROVED** [review]
There's a lot in here that we will need to put into a better shape, but I think it is important enough to get this stake in the ground that I'm mostly holding comment in favor of working it forward. Comments that I have left are nits in that context.

File: build_tools/third_party/triton/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,361 @@
+# Copyright 2023 The OpenXLA Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+set(TRITON_ROOT "${OPENXLA_NVGPU_ROOT_DIR}/third_party/triton/")
```
[the remark this answers is not in the document]
Comment (ezhulenev) [reply build_tools/third_party/triton/CMakeLists.txt:7]: 👍

**Pull request merged**